File: solutions/elixir/collatz-conjecture/1/lib/collatz_conjecture.ex
defmodule CollatzConjecture do
  defguard is_even(input) when input > 0 and rem(input, 2) == 0
  defguard is_odd(input) when input > 0 and rem(input, 2) == 1
  @doc """
  calc/1 takes an integer and returns the number of steps required to get the
  number to 1 when following the rules:
    - if number is odd, multiply with 3 and add 1
    - if number is even, divide by 2
  """
  @spec calc(input :: pos_integer()) :: non_neg_integer()
  def calc(input, steps \\ 0)
  def calc(1, steps), do: steps
  def calc(input, steps) when is_even(input), do: calc(div(input, 2), steps + 1)
  def calc(input, steps) when is_odd(input), do: calc(3*input + 1, steps + 1)
end

File: solutions/python/collatz-conjecture/5/collatz_conjecture.py
def even(number):
    """
    Calculate the number of steps for an even number in the Collatz sequence.

    Parameters:
    number (int): A positive even integer to process.

    Returns:
    int: The result of steps for the even number.
    """
    return steps(number // 2)


def uneven(number):
    """
    Calculate the number of steps for an odd number in the Collatz sequence.

    Parameters:
    number (int): A positive odd integer to process.

    Returns:
    int: The result of steps for the odd number.
    """
    return steps(3 * number + 1)


def steps(number):
    """
    Calculate the number of steps required to reduce the given positive integer
    to 1 according to the rules of the Collatz conjecture.

    The rules are:
    - If the number is even, divide it by 2.
    - If the number is odd, multiply it by 3 and add 1.
    - Repeat the process until the number becomes 1.

    Parameters:
    number (int): A positive integer to be reduced to 1.

    Returns:
    int: The number of steps required to reduce the input number to 1.

    Raises:
    ValueError: If the input number is less than 1 (only positive integers are allowed).
    """
    if number < 1:
        raise ValueError("Only positive integers are allowed")

    if number == 1:
        return 0

    return (uneven(number) if number & 1 else even(number)) + 1
